crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -191,46 +228,507 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
+pub struct StateTransition;
+
+/// Applies manual state transitions using [`NextState<S>`]
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ManualStateTransitions;
+
+fn internal_apply_state_transition<S: States>(world: &mut World, new_state: Option<S>) {
+    match new_state {
+        Some(entered) => {
+            match world.get_resource_mut::<State<S>>() {
+                Some(mut state_resource) => {
+                    if *state_resource != entered {
+                        let exited = mem::replace(&mut state_resource.0, entered.clone());
+                        world.send_event(StateTransitionEvent {
+                            before: exited.clone(),
+                            after: entered.clone(),
+                        });
+                        world
+                            .try_run_schedule(ComputeDependantStates::<S>::default())
+                            .ok();
+
+                        let mut state_transition_schedules =
+                            world.get_resource_or_insert_with(StateTransitionSchedules::default);
+                        state_transition_schedules
+                            .exit_schedules
+                            .insert(OnExit(exited.clone()).intern());
+                        state_transition_schedules.transition_schedules.insert(
+                            OnTransition {
+                                from: exited,
+                                to: entered.clone(),
+                            }
+                            .intern(),
+                        );
+                        state_transition_schedules
+                            .enter_schedules
+                            .insert(OnEnter(entered.clone()).intern());
+                    }
+                }
+                None => {
+                    world.insert_resource(State(entered.clone()));
+                    world
+                        .try_run_schedule(ComputeDependantStates::<S>::default())
+                        .ok();
+
+                    let mut state_transition_schedules =
+                        world.get_resource_or_insert_with(StateTransitionSchedules::default);
+                    state_transition_schedules
+                        .enter_schedules
+                        .insert(OnEnter(entered.clone()).intern());
+                }
+            };
+        }
+        None => {
+            if let Some(resource) = world.remove_resource::<State<S>>() {
+                world
+                    .try_run_schedule(ComputeDependantStates::<S>::default())
+                    .ok();
+
+                let mut state_transition_schedules =
+                    world.get_resource_or_insert_with(StateTransitionSchedules::default);
+                state_transition_schedules
+                    .exit_schedules
+                    .insert(OnExit(resource.0).intern());
+            }
+        }
+    }
+}
```

Comment:
We need either some comments here or function extraction, it's hard to read

Reply:
Will do